File: src/python/boids.py
import pygame
from linalg import Vector3D
import colors
import numpy as np
import utilities

class Boid(utilities.DrawSprite):
  MARGIN = 100 # how close to edges are boids allowed?
  TURN_FACTOR = 2 # how quickly do boids avoid edges?
  MIN_SPEED = 10
  MAX_SPEED = 15
  PROTECTED_RANGE = 10 # the range at which boids get paranoid of collisions with others
  AVOID_FACTOR = 0.1 # how quickly do boids avoid each other in the protected range?
  VISIBLE_RANGE = 50 # this boid will follow boids within this range
  MATCHING_FACTOR = 0.2 # how quickly does this boid follow others?
  CENTERING_FACTOR = 0.005 # how much does this boid stay with others?
  flock = [] # the existing flock of boids

  def __init__(self, bounds):
    # Call the parent class (Sprite) constructor
    pygame.sprite.Sprite.__init__(self)

    # initialize RNG
    rng = np.random.default_rng()

    (self.left, self.right), (self.top, self.bottom), (self.back, self.front) = bounds

    self.areaWidth = self.right - self.left
    self.areaHeight = self.bottom - self.top
    self.areaDepth = self.front - self.back

    x = self.left + self.areaWidth * rng.random()
    y = self.top + self.areaHeight * rng.random()
    z = self.back + self.areaDepth * rng.random()
    vx = 5 * (1 + rng.random())
    vy = 5 * (1 + rng.random())
    vz = 10 * rng.random()

    self.position = Vector3D(x, y, z)
    self.velocity = Vector3D(vx, vy, vz)
    self.closeness = Vector3D.zero()
    self.width = 5
    self.rect = pygame.Rect(self.position.X - self.width / 2, 
                            self.position.Y - self.width / 2,
                            self.width, self.width)
    self.image = pygame.Surface([self.width, self.width], flags=pygame.SRCALPHA)

    Boid.flock.append(self)

  @property
  def color(self):
    return *colors.WHITE, self.position.Z

  def constrain_speed(self):
    speed = abs(self.velocity)
    if speed > Boid.MAX_SPEED:
      self.velocity = (self.velocity * Boid.MAX_SPEED) / speed
    if speed < Boid.MIN_SPEED:
      self.velocity = (self.velocity * Boid.MIN_SPEED) / speed

  def avoid_edges(self):
    if self.position.X < self.left + Boid.MARGIN:
      self.velocity.X += Boid.TURN_FACTOR
    if self.position.X > self.right - Boid.MARGIN:
      self.velocity.X -= Boid.TURN_FACTOR
    if self.position.Y < self.top + Boid.MARGIN:
      self.velocity.Y += Boid.TURN_FACTOR
    if self.position.Y > self.bottom - Boid.MARGIN:
      self.velocity.Y -= Boid.TURN_FACTOR
    if self.position.Z < self.back + Boid.MARGIN:
      self.velocity.Z += Boid.TURN_FACTOR
    if self.position.Z > self.front - 50:
      self.velocity.Z -= Boid.TURN_FACTOR

  def constrain_position(self):
    self.position.X = utilities.constrain(self.position.X, self.left, self.right)
    self.position.Y = utilities.constrain(self.position.Y, self.top, self.bottom)
    self.position.Z = utilities.constrain(self.position.Z, self.back + (self.areaDepth * 0.5), self.front)

  def avoid_other_boids(self):
    self.velocity += self.closeness * Boid.AVOID_FACTOR

  def follow_neighbors(self):
    if self.num_neighbors > 0:
      self.avg_velocity /= self.num_neighbors
      self.avg_position /= self.num_neighbors
      self.velocity += (self.avg_velocity - self.velocity) * Boid.MATCHING_FACTOR
      self.velocity += (self.avg_position - self.position) * Boid.CENTERING_FACTOR

  def fly_with_flock(self):
    self.closeness = Vector3D.zero()
    self.avg_velocity = Vector3D.zero()
    self.avg_position = Vector3D.zero()
    self.num_neighbors = 0
    for other in Boid.flock:
      if self is not other:
        distance = abs(self.position - other.position)
        if distance < Boid.PROTECTED_RANGE:
          self.closeness += self.position - other.position
        if distance < Boid.VISIBLE_RANGE:
          self.avg_velocity += other.velocity
          self.avg_position += other.position
          self.num_neighbors += 1
    self.avoid_other_boids()
    self.follow_neighbors()


  def update(self):
    self.fly_with_flock()
    self.avoid_edges()
    self.constrain_speed()
    self.position += self.velocity
    self.constrain_position()
    self.rect = pygame.Rect(self.position.X - self.width / 2, 
                            self.position.Y - self.width / 2,
                            self.width, self.width)
    self.image.fill(self.color)

File: src/c/boids.c
#include "boids.h"

int main(int argc, char *argv[])
{

    // returns zero on success else non-zero
    if (SDL_Init(SDL_INIT_EVERYTHING) != 0) {
        printf("error initializing SDL: %s\n", SDL_GetError());
    }
    SDL_Window* win = SDL_CreateWindow("Boids", // creates a window
                                    SDL_WINDOWPOS_CENTERED,
                                    SDL_WINDOWPOS_CENTERED,
                                    SCREEN_WIDTH, SCREEN_HEIGHT, 0);

    // triggers the program that controls
    // your graphics hardware and sets flags
    Uint32 render_flags = SDL_RENDERER_ACCELERATED;

    // creates a renderer to render our images
    SDL_Renderer* rend = SDL_CreateRenderer(win, -1, render_flags);

    // creates flock
    Boid flock[N_BOIDS];
    for (int i = 0; i < N_BOIDS; i++)
    {
        initBoid(rend, &flock[i]);
    }

    // controls animation loop
    int close = 0;

    // animation loop
    while (!close) {
        SDL_Event event;

        // Events management
        while (SDL_PollEvent(&event)) {
            switch (event.type) {

            case SDL_QUIT:
                // handling of close button
                close = 1;
                break;

            case SDL_KEYDOWN:
                // keyboard API for key pressed
                switch (event.key.keysym.scancode) {
                case SDL_SCANCODE_ESCAPE:
                    close = 1;
                    break;
                default:
                    break;
                }
            }
        }

        // update flock
        for (int i = 0; i < N_BOIDS; i++)
        {
            updateBoid(&flock[i], flock);
            drawBoid(rend, &flock[i]);
        }

        // clears the screen
        SDL_RenderClear(rend);
        for (int i = 0; i < N_BOIDS; i++)
        {
            SDL_RenderCopy(rend, flock[i].tex, NULL, &flock[i].rect);
        }

        // triggers the double buffers
        // for multiple rendering
        SDL_RenderPresent(rend);

        // calculates to 60 fps
        SDL_Delay(1000 / 60);
    }

    // destroy texture
    for (int i = 0; i < N_BOIDS; i++)
    {
        SDL_DestroyTexture(flock[i].tex);
    }

    // destroy renderer
    SDL_DestroyRenderer(rend);

    // destroy window
    SDL_DestroyWindow(win);
    
    // close SDL
    SDL_Quit();

    return 0;
}

/*
    Place the given Boid somewhere randomly along the screen
 */
void placeBoid(Boid* boid)
{
    boid->position.x = rand() % SCREEN_WIDTH;
    boid->position.y = rand() % SCREEN_HEIGHT;
    boid->position.z = rand() % SCREEN_DEPTH;
    int choice;
    choice = rand();
    boid->velocity.x = ((choice % 2) ? 1 : -1) * (0.5 * MIN_SPEED + (MAX_SPEED - MIN_SPEED) * ((choice % 1000) / 1000.0));
    choice = rand();
    boid->velocity.y = ((choice % 2) ? 1 : -1) * (0.5 * MIN_SPEED + (MAX_SPEED - MIN_SPEED) * ((choice % 1000) / 1000.0));
    choice = rand();
    boid->velocity.z = ((choice % 2) ? 1 : -1) * SEED_DEPTH_SPEED * ((choice % 1000) / 1000.0);
}

/*
    Initialize Boid parameters
 */
int initBoid(SDL_Renderer* rend, Boid* boid)
{
    int ret;

    // place boid somewhere
    placeBoid(boid);

    // create the surface
    boid->surface = SDL_CreateRGBSurfaceWithFormat(0, BOID_SIZE, BOID_SIZE, 32, SDL_PIXELFORMAT_RGBA32);
    ret = SDL_FillRect(boid->surface, NULL, SDL_MapRGBA(boid->surface->format, 255, 255, 255, boid->position.z));

    if (ret)
    {
        return ret;
    }

    // loads image to our graphics hardware memory.
    boid->tex = SDL_CreateTextureFromSurface(rend, boid->surface);

    // clears main-memory
    SDL_FreeSurface(boid->surface);

    // connects our texture with rect to control position
    ret = SDL_QueryTexture(boid->tex, NULL, NULL, &(boid->rect.w), &(boid->rect.h));

    // position Boid
    boid->rect.x = boid->position.x;
    boid->rect.y = boid->position.y;

    return ret;
}

/*
    Make Boid avoid screen edges
 */
void avoidEdges(Boid* boid)
{
    if (boid->position.x < MARGIN)
        boid->velocity.x += TURN_FACTOR;
    if (boid->position.x > SCREEN_WIDTH - MARGIN)
        boid->velocity.x -= TURN_FACTOR;
    if (boid->position.y < MARGIN)
        boid->velocity.y += TURN_FACTOR;
    if (boid->position.y > SCREEN_HEIGHT - MARGIN)
        boid->velocity.y -= TURN_FACTOR;
    if (boid->position.z < MARGIN)
        boid->velocity.z += TURN_FACTOR;
    if (boid->position.z > SCREEN_HEIGHT - FRONT_MARGIN)
        boid->velocity.z -= TURN_FACTOR;
}

/*
    Ensure Boid speed remains within range
 */
void constrainSpeed(Boid* boid)
{
    double speed = sqrt(boid->velocity.x * boid->velocity.x + boid->velocity.y * boid->velocity.y);
    if (speed > MAX_SPEED)
    {
        boid->velocity.x = (boid->velocity.x * MAX_SPEED) / speed; 
        boid->velocity.y = (boid->velocity.y * MAX_SPEED) / speed; 
    }
    if (speed < MIN_SPEED)
    {
        boid->velocity.x = (boid->velocity.x * MIN_SPEED) / speed; 
        boid->velocity.y = (boid->velocity.y * MIN_SPEED) / speed; 
    }
}

/*
    Constrain Boid to always stay within screen
 */
void constrainPosition(Boid* boid)
{
    if (boid->position.x < 0)
        boid->position.x = 0;
    if (boid->position.x > SCREEN_WIDTH)
        boid->position.x = SCREEN_WIDTH;
    if (boid->position.y < 0)
        boid->position.y = 0;
    if (boid->position.y > SCREEN_HEIGHT)
        boid->position.y = SCREEN_HEIGHT;
    if (boid->position.z < MARGIN)
        boid->position.z = MARGIN;
    if (boid->position.z > SCREEN_DEPTH)
        boid->position.z = SCREEN_DEPTH;
}

/*
    Follow neighboring boids
 */
void followNeighbors(Boid* boid)
{
    if (boid->neighbors > 0)
    {
        boid->avgPosition = multiply(boid->avgPosition, 1.0 / boid->neighbors);
        boid->avgVelocity = multiply(boid->avgVelocity, 1.0 / boid->neighbors);
        boid->velocity = add(boid->velocity, multiply(sub(boid->avgVelocity, boid->velocity), MATCHING_FACTOR));
        boid->velocity = add(boid->velocity, multiply(sub(boid->avgPosition, boid->position), CENTERING_FACTOR));
    }
}

/*
    Avoid other boids
 */
void avoidOthers(Boid* boid)
{
    boid->velocity = add(boid->velocity, multiply(boid->closeness, AVOID_FACTOR));
}

/*
    Fly with the flock
 */
void flyWithFlock(Boid* boid, Boid* flock)
{
    zero(&boid->closeness);
    zero(&boid->avgPosition);
    zero(&boid->avgVelocity);
    boid->neighbors = 0;
    for (int i = 0; i < N_BOIDS; i++)
    {
        Boid* other = (flock + i);
        if (boid == other)
            continue;

        Vector3D diff = sub(boid->position, other->position);
        double dist = length(diff);
        if (dist < PROTECTED_RANGE)
            boid->closeness = add(boid->closeness, diff);
        if (dist < VISIBLE_RANGE)
        {
            boid->avgPosition = add(boid->avgPosition, other->position);
            boid->avgVelocity = add(boid->avgVelocity, other->velocity);
            boid->neighbors++;
        }
    }
    avoidOthers(boid);
    followNeighbors(boid);
}

/*
    Flutter depth-wise for nice effect
 */
void flutterDepth(Boid* boid)
{
    int choice = rand();
    boid->velocity.z += (choice % 2 ? 1 : -1) * FLUTTER_SPEED * (choice % DEPTH_FLUTTER);
}

/*
    Update loop for indivual Boid
 */
void updateBoid(Boid* boid, Boid* flock)
{
    flyWithFlock(boid, flock);
    avoidEdges(boid);
    constrainSpeed(boid);
    flutterDepth(boid);
    boid->position = add(boid->position, boid->velocity);
    constrainPosition(boid);
    boid->rect.x = boid->position.x;
    boid->rect.y = boid->position.y;
}

void drawBoid(SDL_Renderer* rend, Boid* boid)
{
    // create the surface
    boid->surface = SDL_CreateRGBSurfaceWithFormat(0, BOID_SIZE, BOID_SIZE, 32, SDL_PIXELFORMAT_RGBA32);
    SDL_FillRect(boid->surface, NULL, SDL_MapRGBA(boid->surface->format, 255, 255, 255, boid->position.z));
    SDL_DestroyTexture(boid->tex);
    boid->tex = SDL_CreateTextureFromSurface(rend, boid->surface);
    // clears main-memory
    SDL_FreeSurface(boid->surface);
}